Enable multi-buffer per channel in EDM

SeanNijjar opened the pull request:
### Ticket
#6300

### Problem description
The EDM currently only supports one buffer per channel which can limit performance for CCL ops. 

### What's changed
Add multi-buffering per EDM channel

Adds the option to add a multiple buffers (e.g. double buffered) per EDM
channel. This is useful for improving performance of CCL operation.

To simplify the worker <-> EDM interface to allow a kernel to
automatically support multi-buffered channels, new adapter components
are added:
 - WorkerToEdmReader: for a worker pulling data from EDM
 - WorkerToEdmSender: for a worker pushing data to the EDM

These hide details such as buffer offsets in the channel and any other
details that may only be relevant to the EDM. Additionally, their use
encapsulates the worker <-> EDM data movement protocol, allowing future
low level changes to buffer layouts and allocations on the EDM without
requiring worker kernel changes.

As an a coinciding required step to enable this functionality the EDM
channel count limit has been lifted to unlimited (limited only to as
many buffers can fit into L1). This provides additional flexibility for
op writers and let's the erisc_info::channels to be shrunk back to
single entry.

Note that this commit only adds this feature, but does not yet enable it
for CCL ops.

Multi-buffering is a useful and important component of improved performance for CCLs (for example, for measured  allgathers operations on a separate branch when using this feature, we see a "free" 15-25% cycle count reduction - seemingly across the board).

### Checklist
- [x] Post commit CI passes: https://github.com/tenstorrent/tt-metal/actions/runs/10441905110 
- [x] t3000 frequent: https://github.com/tenstorrent/tt-metal/actions/runs/10441933102
- [x] t3000 unit test: https://github.com/tenstorrent/tt-metal/actions/runs/10442795490
  - Same failures as main on same commit: https://github.com/tenstorrent/tt-metal/actions/runs/10441588465
  - Personal branch was also green on earlier rebase off main: https://github.com/tenstorrent/tt-metal/actions/runs/10388731388
- [x] t3000 model perf: https://github.com/tenstorrent/tt-metal/actions/runs/10441913051
- [x] t3000 demos: https://github.com/tenstorrent/tt-metal/actions/runs/10442938382
  - Same failures as main on same commit: https://github.com/tenstorrent/tt-metal/actions/runs/10440099925
- [x] Blackhole Post commit (if applicable)
- [x] Model regression CI testing passes (if applicable) 
- [x] New/Existing tests provide coverage for changes

# Perf Numbers After
- Reduce scatter 
[reduce_scatter_after.csv](https://github.com/user-attachments/files/16604011/reduce_scatter_after.csv)
- All gather post commit :
[all_gather_post_commit_after.csv](https://github.com/user-attachments/files/16604922/all_gather_post_commit_after.csv)
- All gather nightly subset:
[all_gather_nightly_subset_after.csv](https://github.com/user-attachments/files/16605116/all_gather_nightly_subset_after.csv)

# Perf numbers before
- Reduce scatter 
[perf_reduce_scatter_post_commit_before.csv](https://github.com/user-attachments/files/16607726/perf_reduce_scatter_post_commit_before.csv)
- All gather post commit 
[perf_all_gather_post_commit_before.csv](https://github.com/user-attachments/files/16607727/perf_all_gather_post_commit_before.csv)


# Plotted Delta (Reduce scatter)
<img width="1654" alt="image" src="https://github.com/user-attachments/assets/52eddbe4-9350-4590-b6e3-33cfd9bd7479">

Mostly massive improvement with a few cases seeing degradation:

- Interleaved: Input [32,2048], fp16, output [32,256], tile, dim=3, DRAM 
- Interleaved: Input [32,2048], bfp8, output [32,256], tile, dim=3, DRAM
- Interleaved: Input [32,2048], fp16, output [32,256], tile, dim=3, L1 
- Interleaved: Input [32,2048], bfp8, output [32,256], tile, dim=3, L1

@johanna-rock-tt @cglagovich @djordje-tt - is this config used anywhere? Based on the overall improvements, and the fact that this work unblocks quite a bit of future optimization, I'm tempted to merge this, despite the degradation in a couple cases, and file an issue with the specific case and before/after numbers to recover perf later. (including exact commits for reproducibility)

# Plotted Delta (All Gather):
All gather perf delta is mostly neutral (expected) with a few swings:
<img width="1667" alt="image" src="https://github.com/user-attachments/assets/05527acb-f3f0-4688-a528-2ad335d28524">

 All degradations are with line all-gather and we currently don't have a good understanding of line-all-gather performance variability. Similar to reduce scatter, I'd like to go ahead and perform a similar exercise with the line all-gathers (which are primarily used in galaxy configs and in the process of being brought up/tightened).

SeanNijjar commented:
Some pipelines are failing but the failing jobs seem to have consistent signatures as on main.
(From model demos) https://github.com/tenstorrent/tt-metal/actions/runs/10389210030/job/28767748641 (@djordje-tt )

TT-billteng commented on .github/workflows/ttnn-post-commit.yaml:
```diff
@@ -60,6 +60,8 @@ jobs:
           - name: ttnn examples and cpp tests
             cmd: ./build/test/ttnn/unit_tests_ttnn && ./tests/scripts/run_ttnn_examples.sh
+          - name: ttnn ccl cpp unit tests
+            cmd: ./build/test/ttnn/unit_tests_ttnn_ccl
```
please add this to one of the scripts called by `run_test.sh` so that devs who want to run locally pick it up also

cglagovichTT commented:
@SeanNijjar Llama doesn't have any of those small reduce-scatter shapes which saw the perf decrease

tt-rkim commented on .github/workflows/ttnn-post-commit.yaml:
```diff
@@ -60,6 +60,8 @@ jobs:
             fast_runtime_mode_off: true
           - name: ttnn examples and cpp tests
             cmd: ./build/test/ttnn/unit_tests_ttnn && ./tests/scripts/run_ttnn_examples.sh
+          - name: ttnn ccl cpp unit tests
```
This workflow does not run T3000 unit tests. Could you clarify your intention?

SeanNijjar replied:
That's fine. These tests are intended to be able to run on n300.

tt-rkim replied:
Sounds good. Did you verify that it actually runs on post commit CI?

SeanNijjar replied:
Yessir:
<img width="248" alt="image" src="https://github.com/user-attachments/assets/61426b0d-f704-41ee-aab5-8c4aee85f63c">

tt-rkim replied:
<img width="621" alt="Screenshot 2024-08-16 at 10 45 04 AM" src="https://github.com/user-attachments/assets/fecaa636-a5b6-4383-b372-f5bda156b93c">

is this what we want? This is `ttnn-unit-tests (wormhole_b0, N300) / ttnn ccl cpp unit tests wormhole_b0 N300`

SeanNijjar replied:
Unfortunately yes. 

I wanted to enable this as a slot to add further unit tests and a placeholder until that issue seen on CI runs is resolved and to keep the build working... though now I think of it, I guess the test should still be building regardless of these new test bucket's presence. 

Some further context for my reasoning:
This test - the same test and code - with the 15 different cases, has some reliability (hang) issues on CI for an as yet undetermined reason. However, the test test is, so far, stable across every board and machine I've tried (in total equaling billions of messages and thousands of test invocations across 5 boards). Additionally, the main component tested by these tests (EDM), was tested stable against millions of all-gather op invocations.

SeanNijjar committed:
#6300: Add multi-buffering per EDM channel

Adds the option to add a multiple buffers (e.g. double buffered) per EDM
channel. This is useful for improving performance of CCL operation.

To simplify the worker <-> EDM interface to allow a kernel to
automatically support multi-buffered channels, new adapter components
are added:
 - WorkerToEdmReader: for a worker pulling data from EDM
 - WorkerToEdmSender: for a worker pushing data to the EDM

These hide details such as buffer offsets in the channel and any other
details that may only be relevant to the EDM. Additionally, their use
encapsulates the worker <-> EDM data movement protocol, allowing future
low level changes to buffer layouts and allocations on the EDM without
requiring worker kernel changes.

As an a coinciding required step to enable this functionality the EDM
channel count limit has been lifted to unlimited (limited only to as
many buffers can fit into L1). This provides additional flexibility for
op writers and let's the `erisc_info::channels` to be shrunk back to
single entry.

Note that this commit only adds this feature, but does not yet enable it
for CCL ops.
